<launch>
  <arg name="manager" default="zvision_lidar_nodelet_manager" />
  <arg name="model" default="ML30SA1" />
  <arg name="device_ip" default="192.168.10.108" />
  <arg name="udp_port" default="2368" />
  <arg name="angle_path" default="" />
  <arg name="pcap" default="" />
  <arg name="publisher_pointcloud_type_xyzi" default="false" />
  <arg name="use_blue_gold_color_scheme" default="false" />
  <arg name="publisher_pointcloud_type_xyzirt" default="true" />
  <arg name="dynamic_reconfigure_server" default="true" />
  <arg name="publisher_frame_info" default="true" />
  <arg name="use_lidar_line_id" default="false" />
  <arg name="use_pointcloud_status_diagnose" default="true" />
  <arg name="statistic_frame_num" default="100" />
  <arg name="roi_sample_lines" default="10" />
  <arg name="roi_interval" default="10" />
  <arg name="min_roi_pointnum" default="60" />
  <arg name="z_height" default="-0.58" />
  <arg name="roi_z_diff_threshold" default="0.1" />
  <arg name="error_rate_threshold" default="0.5" />
  <!-- start nodelet manager -->
  <node pkg="nodelet" type="nodelet" name="$(arg manager)" args="manager" output="screen"/>

  <!-- load driver nodelet into it (udp packets)-->
  <node pkg="nodelet" type="nodelet" name="$(arg manager)_driver"
        args="load zvision_lidar_driver/DriverNodelet $(arg manager)" output="screen">
    <param name="model" value="$(arg model)"/>
    <param name="device_ip" value="$(arg device_ip)" />
    <param name="udp_port" value="$(arg udp_port)" />
	<param name="pcap" value="$(arg pcap)" />
  <param name="dynamic_reconfigure_server" value="$(arg dynamic_reconfigure_server)" />
  </node>

  <!-- load driver nodelet into it (udp packet to pointcloud)-->
  <node pkg="nodelet" type="nodelet" name="$(arg manager)_cloud"
        args="load zvision_lidar_pointcloud/CloudNodelet $(arg manager)" output="screen">
    <param name="model" value="$(arg model)"/>
    <param name="angle_path" value="$(arg angle_path)" />
    <param name="device_ip" value="$(arg device_ip)" />
    <!-- <param name="downsample_type" value="downsample_voxel" /> -->
    <!-- <param name="downsample_type" value="downsample_line" /> -->
    <!-- <param name="downsample_type" value="downsample_cfg_file" /> -->
    <!-- <param name="downsample_type" value="none" /> -->
    <param name="downsample_cfg_path" value="$(find zvision_lidar_pointcloud)/data/Downsample_ML30S_A1_1_2.txt" />
    <param name="line_sample" value="2" />
    <param name="voxel_leaf_size" value="0.5" />
    <param name="x_tra" value="0.0" />
    <param name="y_tra" value="0.0" />
    <param name="z_tra" value="0.0" />
    <param name="x_rot" value="0.0" />
    <param name="y_rot" value="0.0" />
    <param name="z_rot" value="0.0" />
    <param name="use_lidar_line_id" value="$(arg use_lidar_line_id)" />
    <param name="use_lidar_time" value="true" />
    <param name="use_outlier_removal" value="false"/> 
    <param name="outlier_th" value="0.25"/>
    <param name="publisher_pointcloud_type_xyzi" value="$(arg publisher_pointcloud_type_xyzi)" />
    <param name="use_blue_gold_color_scheme" value="$(arg use_blue_gold_color_scheme)"/>
    <param name="publisher_pointcloud_type_xyzirt" value="$(arg publisher_pointcloud_type_xyzirt)" />
    <param name="dynamic_reconfigure_server" value="$(arg dynamic_reconfigure_server)" />
    <param name="publisher_frame_info" value="$(arg publisher_frame_info)" />
    <param name="use_pointcloud_status_diagnose" value="$(arg use_pointcloud_status_diagnose)"/>
    <param name="statistic_frame_num" value="$(arg statistic_frame_num)"/>
    <param name="roi_sample_lines" value="$(arg roi_sample_lines)"/>
    <param name="roi_interval" value="$(arg roi_interval)"/>
    <param name="min_roi_pointnum" value="$(arg min_roi_pointnum)"/>
    <param name="z_height" value="$(arg z_height)"/>
    <param name="roi_z_diff_threshold" value="$(arg roi_z_diff_threshold)"/>
    <param name="error_rate_threshold" value="$(arg error_rate_threshold)"/>
  </node>
  
  <node if="$(arg use_blue_gold_color_scheme)" name="rviz" pkg="rviz" type="rviz"  args="-d $(find zvision_lidar_pointcloud)/rviz_cfg/zvision_lidar_colored.rviz" />
  <group unless="$(arg use_blue_gold_color_scheme)">
    <node if="$(arg publisher_pointcloud_type_xyzirt)" name="rviz" pkg="rviz" type="rviz"  args="-d $(find zvision_lidar_pointcloud)/rviz_cfg/zvision_lidar_xyzirt.rviz" />
    <node unless="$(arg publisher_pointcloud_type_xyzirt)" name="rviz" pkg="rviz" type="rviz"  args="-d $(find zvision_lidar_pointcloud)/rviz_cfg/zvision_lidar.rviz" />
  </group>

</launch>
